src/mbgl/renderer/renderer_impl.cpp

```diff
@@ -369,6 +374,47 @@ void Renderer::Impl::render(const UpdateParameters& updateParameters) {
         parameters.frameHistory.upload(parameters.context, 0);
     }
 
+    // - PREPARE + CLIP ------------------------------------------------------------------------------
+    {
+        MBGL_DEBUG_GROUP(parameters.context, "clip");
+
+        // Update all clipping IDs.
+        for (const auto& entry : renderSources) {
+            if (entry.second->isEnabled()) {
+                entry.second->startRender(parameters);
+            }
+        }
+    }
+
+    // - 3D PASS -------------------------------------------------------------------------------------
+    // Renders any 3D layers bottom-to-top to unique FBOs with texture attachments, but share the same
+    // depth rbo between them.
+    if (parameters.staticData.has3D) {
+        MBGL_DEBUG_GROUP(parameters.context, "3d");
+        parameters.backend.bind();
+        parameters.pass = RenderPass::Pass3D;
+
+        const auto size = parameters.context.viewport.getCurrentValue().size;
```

Comment:
`parameters.context` contains the value of the current OpenGL state (unless it is dirty). Instead of using this value, we should use the canonical value from `parameters.state.getSize()`, since the OpenGL state's value could be dirty or set to some other value if we don't bind the backend first (see above).

Reply:
@kkaefer interesting — when I change this to `parameters.state.getSize()` here (renderbuffer creation) and in framebuffer/texture creation

https://github.com/mapbox/mapbox-gl-native/blob/0efb0b75602b4ae177e7faf8869c4f82866bf7f1/src/mbgl/renderer/renderer_impl.cpp#L392-L404
https://github.com/mapbox/mapbox-gl-native/blob/0efb0b75602b4ae177e7faf8869c4f82866bf7f1/src/mbgl/renderer/layers/render_fill_extrusion_layer.cpp#L54-L59

on my iPhone, it appears you're right that this value is different, but that the dimensions of `parameters.state.getSize()` are half those of the others (non-retina). If I change them, fill-extrusion layers degrade to non-retina quality…